design/ceph/object/object-store-backend.md

```diff
@@ -0,0 +1,103 @@
+---
+title: Object Store Backend
+target-version: release-1.10
+---
+
+# Overview
+
+Ceph RADOS Gateway is undergoing a major overhaul with the Zipper project. The goal of Project
+Zipper is to create an API that can be used to separate the RGW top end (represented by Swift and
+S3) from bottom end providers (represented currently by RADOS, and eventually by Cloud, File, and
+Memory backends). This should allow the front ends to access any implemented back end
+transparently.
+
+The secondary objective is to allow multiple layers of these API providers to stack up, allowing
+transformations of operations (such as redirection to different backends), or caching layers that
+cut across multiple backends.
+
+The primary Store is the current RadosStore, which provides access to a Ceph cluster via RADOS.
+However, new Stores are possible that store the data in any desired platform. One such Store, called
+DBStore, has been developed that stores data in SQL, and specifically in a local SQLite database.
+
+## Goals
+
+Rook-Ceph has a tremendous support for the Ceph Object Store and we want to leverage that.
+Especially the various controllers that are reconciling ObjectStore, Multisite configurations and
+Bucket provisioning.
+Mainly we are looking at providing the next generation of RGW via Rook.
+
+# Proposal details
+
+The `CephObjectStore` API will have a few details added to its spec such as an image so that it can
+be deployed without a Ceph cluster.
+
+```go
+type ObjectStoreSpec struct {
+...
+	// Image is the name of the image to use to deploy the object store
+	Image string `json:"image"`
+...
+}
+```
+
+Also, we need to expose a new type for the backend to use:
+
+```go
+type GatewaySpec struct {
+...
+	// Backend refers to the backend to use for the object store
+	// +optional
+	Backend *ObjectStoreBackend `json:"backend,omitempty"`
+...
+}
+```
+
+Where the `ObjectStoreBackend` is represented as such:
+
+```go
+type ObjectStoreBackend struct {
+	// The name of the backend
+	// +optional
+	// +kubebuilder:validation:Enum={"rados","sqlite"}
+	Name string `json:"name,omitempty"`
+
+    // Config is the configuration for the backend
+	// +optional
+    Config map[string]string `json:"config,omitempty"`
+}
+```
```

Comment:
Above, it seems that multiple backends can be used simultaneously. Would it be better to do this with structs for each backend? 

```go
ObjectStoreBackends struct {
  Rados {
    RadosConfig { ... }
  }
  SQLite {
    SQLiteConfig { ... }
  }
}
```

Reply:
Before I saw this comment, I was also going to suggest specific sections for each backend to give stronger typing for the options.

Reply:
Yaa this makes more sense to me as well than a generic config map to hold the info

Reply:
I don't think multiple backends can be used simultaneously, we want to deploy A given backend and manage it, regardless of the number of "instances" to run.
I'm not sure what kind of strong typing we need but having a type for each backend seems a bit of an endless run. Should we change the API each time we have a new backend?
The current approach allows us to transparently add more backends without touching the API.

Reply:
Is it a plugin model where new backends could work without any implementation change in Rook? If so, then the generic API makes sense. But if we anyway need to implement specific backends in rook, then we could touch the API at the same time.

Reply:
It can be both since they are external vendors but work coming out of the ceph project too.

Reply:
You will be able to use multiple backends at the same time in RGW, although you cannot right now.

Reply:
Will the client decide which backend to consume in that case?

Reply:
Or is that more of a case for caching tiering? Otherwise, at a glance it would seem unusual for the client to choose the backing store.

Reply:
One of the use cases is tiering, yes, another one is to spit the namespace so portions of it are on one store (say local Ceph), and portions of it are on another store (say, remote S3).  I don't know how that should look in rook, sorry.  I've never used rook.
Most common use-case will very likely be a single, local store, but it's quite possible users will want multiples.

Reply:
Makes sense, there will be various use cases for multiple stores. My main point is that the admin setting up zipper will determine the store configuration.